File: agents/hal/TcpServer.cpp
/*
 * Copyright 2016 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <errno.h>

#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <dirent.h>

#include <netdb.h>
#include <netinet/in.h>

#include <sys/types.h>
#include <sys/socket.h>

#include <utils/RefBase.h>

#include <iostream>
#include <sstream>

#include "test/vts/agents/hal/proto/AndroidSystemControlMessage.pb.h"
#include "test/vts/sysfuzzer/common/proto/InterfaceSpecificationMessage.pb.h"
#include "BinderClient.h"
#include "RequestHandler.h"

using namespace std;

namespace android {
namespace vts {

const static int kTcpPort = 5001;


// Starts to run a TCP server (foreground).
int StartTcpServer(const char* fuzzer_path32, const char* fuzzer_path64,
                   const char* spec_dir_path) {
  int sockfd, newsockfd;
  socklen_t clilen;
  struct sockaddr_in serv_addr, cli_addr;

  sockfd = socket(AF_INET, SOCK_STREAM, 0);
  if (sockfd < 0) {
    cerr << "Can't open the socket." << endl;
    return -1;
  }

  // Initialize socket structure
  bzero((char*) &serv_addr, sizeof(serv_addr));

  serv_addr.sin_family = AF_INET;
  serv_addr.sin_addr.s_addr = INADDR_ANY;
  serv_addr.sin_port = htons(kTcpPort);

  // Now bind the host address using bind() call.
  if (::bind(sockfd, (struct sockaddr*) &serv_addr, sizeof(serv_addr)) == -1) {
    cerr << "Binding failed." << endl;
    return -1;
  }

  while (true) {
    // Now start listening for the clients, here process will go in sleep mode
    // and will wait for the incoming connection
    listen(sockfd, 5);
    clilen = sizeof(cli_addr);

    // Accept actual connection from the client
    newsockfd = ::accept(sockfd, (struct sockaddr*) &cli_addr, &clilen);
    if (newsockfd < 0) {
      cerr << "Accept failed" << endl;
      return -1;
    }

    cout << "New session" << endl;
    pid_t pid = fork();
    if (pid == 0) {  // child
      AgentRequestHandler handler;
      exit(handler.StartSession(newsockfd, fuzzer_path32, fuzzer_path64,
                                spec_dir_path));
    } else if (pid < 0){
      cerr << "can't fork a child process to handle a session." << endl;
      return -1;
    }
  }
  return 0;
}

}  // namespace vts
}  // namespace android

File: runners/host/tcp_server/tcp_server.py
#!/usr/bin/env python
#
# Copyright 2016 - The Android Open Source Project
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import SocketServer
import logging
import threading
from socket import error as socket_error
from vts.runners.host.errors import TcpServerCreationError
from vts.runners.host.errors import TcpServerShutdownError

_functions = dict()  # Dictionary to hold function pointers

class TCPRequestHandler(SocketServer.BaseRequestHandler):
    """The request handler class for our server.

    Attributes:
        _error_reponse: The response message sent to client when the function
            is not registered in the dictionary.
        _success_response: Message sent when the function is registered with
            the dictionary.
    """

    _error_response = "Error: Function not registered."
    _success_response = "Success"

    def handle(self):
        """Receives requests from clients.

        This function receives the request from the client and sends the text
        by converting them to upper case characters.
        """
        # self.request is the TCP socket connected to the client
        data = self.request.recv(1024).strip()

        logging.info('client address: ', self.client_address[0])
        logging.info('data: ', data)
        logging.info('Current Thread: ', threading.current_thread())

        if data in _functions:
            _functions[data]()  # call the function pointer
            response = self._success_response
        else:
            response = self._error_response

        self.request.sendall(response)  # send the response back to client


class ThreadedTCPServer(object):
    """This class creates TCPServer in separate thread.

    Attributes:
        _server:   an instance of SocketServer.TCPServer.
        _port_used: this variable maintains the port number used in creating
                  the server connection.
        _IP_address: variable to hold the IP Address of the host.
        _HOST: IP Address to which initial connection is made.
    """
    _server = None
    _port_used = 0  # Port 0 means to select an arbitrary unused port
    _IP_address = ""  # Used to store the IP address for the server
    _HOST = "localhost"  # IP address to which initial connection is made

    def RegisterCallback(self, func_id, callback_func):
        """To register the function pointer.

        This method registers the callback for the _functions by storing the
        function pointer in set.

        Args:
            func_id: Refers to the func_id of function pointers that is maintained in
                the dictionary.
            callback_func:  Refers to the callback_func that we need to
                            register.

        Returns:
            False if func_id is already present in the dictionary, func_id is
            None or callback_func is None; else returns True after inserting
            the key-func_id and value-callback_func.
        """

        if func_id is None or func_id in _functions or callback_func is None:
            return False
        else:
            _functions[func_id] = callback_func  # update the dictionary
            return True

    def UnregisterCallback(self, func_id):
        """Removes the function pointer from the dict corresponding to the key.

        Args:
            func_id: The function id against which the function pointer
                is stored.

        Returns:
            Returns false if func_id is not present in the dictionary -
                _functions, else return True after removing it from dict.
        """

        if func_id is None or func_id not in _functions:
            return False
        else:
            _functions.pop(func_id, None)  #  Remove the key from dictionary
            return True

    def Start(self, port=0):
        """This function starts the TCP server.

        Args:
            port: The port at which connection will be made. Default value
                  is zero, in which case a free port will be chosen
                  automatically.

        Raises:
            TcpServerCreationError: Error occurred while starting server.
        """
        # Start the sever
        try:
            self._server = SocketServer.TCPServer((self._HOST, port),
                                                  TCPRequestHandler)
            self._IP_address, self._port_used = self._server.server_address

            # Start a thread with the server -- that thread will then start one
            # more thread for each request
            server_thread = threading.Thread(target = self._server.serve_forever)

            # Exit the server thread when the main thread terminates
            server_thread.daemon = True
            server_thread.start()
            logging.info('Server loop running in thread: %s', server_thread.name)

        except (RuntimeError, IOError, socket_error) as e:
            logging.exception(e)
            raise TcpServerCreationError('TcpServerCreationError occurred.')

    def Stop(self):
        """This function calls stop server to stop the server instance."""

        # stop the server instance
        self._server.shutdown
        self._server.server_close()

    def GetIPAddress(self):
        """Returns the IP Address used in creating the server."""
        return self._IP_address

    def GetPortUsed(self):
        """Returns the port number used in creating the server."""
        return self._port_used
